File: src/main/java/com/abysalto/middleware/service/ProductMapper.java
package com.abysalto.middleware.service;

import com.abysalto.middleware.domain.Product;
import com.abysalto.middleware.dto.ProductDetailDto;
import com.abysalto.middleware.dto.ProductSummaryDto;
import com.abysalto.middleware.util.TextUtils;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

/**
 * Maps internal {@link Product} domain objects to the API DTOs, applying the short-description
 * truncation for summaries.
 */
@Component
public class ProductMapper {

    private final int descriptionMaxLength;

    public ProductMapper(@Value("${middleware.summary.description-max-length:100}") int descriptionMaxLength) {
        this.descriptionMaxLength = descriptionMaxLength;
    }

    public ProductSummaryDto toSummary(Product p) {
        return new ProductSummaryDto(
                p.thumbnail(),
                p.title(),
                p.price(),
                TextUtils.truncate(p.description(), descriptionMaxLength)
        );
    }

    public ProductDetailDto toDetail(Product p) {
        return new ProductDetailDto(
                p.id(),
                p.title(),
                p.description(),
                p.category(),
                p.price(),
                p.discountPercentage(),
                p.rating(),
                p.stock(),
                p.tags(),
                p.brand(),
                p.sku(),
                p.weight(),
                p.dimensions(),
                p.warrantyInformation(),
                p.shippingInformation(),
                p.availabilityStatus(),
                p.returnPolicy(),
                p.minimumOrderQuantity(),
                p.thumbnail(),
                p.images(),
                p.reviews(),
                p.meta()
        );
    }
}

File: dotnet/src/Middleware.Core/Services/ProductMapper.cs
using Microsoft.Extensions.Options;
using Middleware.Core.Common;
using Middleware.Core.Domain;
using Middleware.Core.Dtos;
using Middleware.Core.Options;

namespace Middleware.Core.Services;

/// <summary>
/// Maps internal <see cref="Product"/> domain objects to the API DTOs, applying the short-description
/// truncation for summaries.
/// </summary>
public sealed class ProductMapper
{
    private readonly int _descriptionMaxLength;

    public ProductMapper(IOptions<SummaryOptions> options)
    {
        _descriptionMaxLength = options.Value.DescriptionMaxLength;
    }

    public ProductSummaryDto ToSummary(Product p) =>
        new(
            p.Thumbnail,
            p.Title,
            p.Price,
            TextUtils.Truncate(p.Description, _descriptionMaxLength));

    public ProductDetailDto ToDetail(Product p) =>
        new(
            p.Id,
            p.Title,
            p.Description,
            p.Category,
            p.Price,
            p.DiscountPercentage,
            p.Rating,
            p.Stock,
            p.Tags,
            p.Brand,
            p.Sku,
            p.Weight,
            p.Dimensions,
            p.WarrantyInformation,
            p.ShippingInformation,
            p.AvailabilityStatus,
            p.ReturnPolicy,
            p.MinimumOrderQuantity,
            p.Thumbnail,
            p.Images,
            p.Reviews,
            p.Meta);
}
